app/src/main/java/com/example/tasky/authentication/presentation/login/viewmodel/LoginViewModel.kt

```diff
@@ -1,14 +1,103 @@
 package com.example.tasky.authentication.presentation.login.viewmodel
 
-class LoginViewModel {
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.tasky.authentication.data.remote.LoginRepository
+import com.example.tasky.authentication.domain.usecase.LoginUseCase
+import com.example.tasky.core.data.Resource
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class LoginViewModel @Inject constructor(
+    private val loginUseCase: LoginUseCase,
+    private val loginRepository: LoginRepository
+) : ViewModel() {
+    var state by mutableStateOf(LoginState())
+        private set
+
+    private var loginJob: Job? = null
+
+    fun onEvent(event: LoginEvent) {
+        when (event) {
+            is LoginEvent.OnEmailValueChanged -> {
+                val isValid = loginUseCase.validateEmailUseCase(event.emailValue)
+                state = state.copy(isValidEmail = isValid, emailValue = event.emailValue)
+            }
+
+            is LoginEvent.OnPasswordValueChanged -> {
+                state = state.copy(passwordValue = event.passwordValue)
+            }
+
+            is LoginEvent.OnShowPasswordValueChanged -> {
+                state = state.copy(showPassword = !state.showPassword)
+            }
+
+            LoginEvent.OnLoginClicked -> {
+                if (!state.isValidEmail) {
+                    state = state.copy(errorMessage = "Invalid email")
+                    return
+                }
+
+                if (loginJob != null) return
+
+                state = state.copy(isLoading = true)
+                loginJob = viewModelScope.launch {
```

Comment:
Why do you need to keep track of the login job? Doesn't seem necessary to me.
If it is to avoid triggering login twice, why don't you rather just add a check `if(state.isLoading) return`? That's a value you already have anyways

Reply:
Yes, thats a better approach